src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,439 @@
+      case _ : BundleType | _ : VectorType =>
+        val newPrefix = prefix + newName + NewLowerTypes.delim
+        val isVecField = oldField.tpe.isInstanceOf[VectorType]
+        val fields = getFields(oldField.tpe)
+        val fieldsWithCorrectOrientation = fields.map(f => f.copy(flip = Utils.times(f.flip, oldField.flip)))
+        val children = fieldsWithCorrectOrientation.flatMap { f =>
+          destruct(newPrefix, RefParentRef(oldRef), f, isVecField, rename.flatMap(_.children.get(f.name)))
+        }
+        // the bundle/vec reference refers to all children
+        children.map{ case(c, r) => (c, r :+ oldRef) }
+    }
+  }
+
+  private case class RenameNode(name: String, children: Map[String, RenameNode])
+
+  /** Implements the core functionality of the old Uniquify pass: rename bundle fields and top-level references
+    * where necessary in order to avoid name clashes when lowering aggregate type with the `_` delimiter.
+    * We don't actually do the rename here but just calculate a rename tree. */
+  private def uniquify(ref: Field, namespace: Namespace): (Option[RenameNode], Seq[String]) = ref.tpe match {
+    case BundleType(fields) =>
+      // we rename bottom-up
+      val localNamespace = new Namespace() ++ fields.map(_.name)
+      val renamedFields = fields.map(f => uniquify(f, localNamespace))
+
+      // Need leading _ for findValidPrefix, it doesn't add _ for checks
+      val renamedFieldNames = renamedFields.flatMap(_._2)
+      val suffixNames: Seq[String] = renamedFieldNames.map(f => NewLowerTypes.delim + f)
+      val prefix = Uniquify.findValidPrefix(ref.name, suffixNames, namespace)
+      // We added f.name in previous map, delete if we change it
+      val renamed = prefix != ref.name
+      if (renamed) {
+        namespace -= ref.name
+        namespace += prefix
+      }
+      val suffixes = renamedFieldNames.map(f => prefix + NewLowerTypes.delim + f)
+
+      val anyChildRenamed = renamedFields.exists(_._1.isDefined)
+      val rename = if(renamed || anyChildRenamed){
+        val children = renamedFields.map(_._1).zip(fields).collect{ case (Some(r), f) => f.name -> r }.toMap
+        Some(RenameNode(prefix, children))
+      } else { None }
+
+      (rename, suffixes :+ prefix)
+    case v : VectorType=>
+      // if Vecs are to be lowered, we can just treat them like a bundle
+      uniquify(ref.copy(tpe = vecToBundle(v)), namespace)
+    case _ : GroundType => (None, List(ref.name))
+    case UnknownType => throw new RuntimeException(s"Cannot uniquify field of unknown type: $ref")
+  }
+
+  private def getFields(tpe: Type): Seq[Field] = tpe match {
+    case BundleType(fields) => fields
+    case v : VectorType => vecToBundle(v).fields
+  }
+
+  private def vecToBundle(v: VectorType): BundleType = {
+    BundleType(( 0 until v.size).map(i => Field(i.toString, Default, v.tpe)))
+  }
+
+  private trait ParentRef { def ref(name: String, asVecField: Boolean = false): ReferenceTarget }
```

Comment:
Please add some explanation of why these are necessary, and what they represent.